### PR TITLE
Add unit-tests for daml2ts

### PR DESCRIPTION
**Pull request by shayne-fletcher**
A unit-test harness for daml2ts and some preliminary tests.

### CODE REVIEW

File: language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        , "template Grover"
+        , "  with puppeteer : Party"
+        , "  where"
+        , "    signatory puppeteer"
+        , "    choice Grover_GoSuper: ContractId Grover"
+        , "      controller puppeteer"
+        , "      do"
+        , "        return self"
+        ]
+      writeFileUTF8 (grover </> "daml.yaml") $ unlines
```
Comment (hurryabit) [inline language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs]: I wonder if we should finally invest in some shared helpers to create all these `daml.yaml` files and the directories containing them. The packaging tests are already full of these repetitions...

File: language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        , "exposed-modules: [Elmo]"
+        , "dependencies:"
+        , "  - daml-prim"
+        , "  - daml-stdlib"
+        ]
+      buildProject elmo ["-o", ".daml" </> "dist" </> "elmo-1.0.dar"]
+      assertBool "elmo-1.0.dar was not created." =<< doesFileExist elmoDar
+      step "Generating TypeScript of 'grover' and 'elmo'..."
+      (exitCode, _, _) <- readProcessWithExitCode daml2ts ([groverDar, elmoDar] ++ ["-o", elmoTs]) ""
+      assertBool "A name collision error was expected." (exitCode /= ExitSuccess)
```
[the remark this answers is not in the document]
Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs]: I can do that.

File: language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        , "  - daml-stdlib"
+        , "  - " <> groverDar
+        ]
+      buildProject charliesRestaurant []
+      assertBool "charlies-restaurant-1.0.dar was not created." =<< doesFileExist charliesRestaurantDar
+      step "Generating TypeScript of 'charlies-restaurant'..."
+      daml2tsProject [charliesRestaurantDar] charliesRestaurantTs
+      assertBool "'CharliesRestaurant.ts' was not created." =<< doesPathExist (charliesRestaurantTs </> "charlies-restaurant-1.0" </> "CharliesRestaurant.ts")
+      assertBool "'packageId.ts' was not created." =<< doesFileExist (charliesRestaurantTs </> "charlies-restaurant-1.0" </> "packageId.ts")
+      assertBool "'Grover.ts' was created." =<< not <$> doesPathExist (charliesRestaurantTs </> "grover-1.0" </> "Grover.ts")
```
[the remark this answers is not in the document]
Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs]: 😄

File: language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+      daml2tsProject [charliesRestaurantDar, groverDar] charliesRestaurantTs
```
[the remark this answers is not in the document]
Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs]: Ok!

File: language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+      assertBool "'Grover.ts' was not created." =<< doesPathExist (charliesRestaurantTs </> "grover-1.0" </> "Grover.ts")
+      assertBool "'packageId.ts' was not created." =<< doesFileExist (charliesRestaurantTs </> "grover-1.0" </> "packageId.ts")
```
[the remark this answers is not in the document]
Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs]: Yes. The tests will evolve with the improved implementation.

File: language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs (hunk 1 above)
Comment (shayne-fletcher) [reply language-support/ts/codegen/tests/src/DA/Test/Daml2Ts.hs]: Yes, there's a mini-library emerging. Maybe not today but soon.

**Comment by shayne-fletcher** [comment]
> Idea for future improvements: call `tsc` in `daml2tsProject` once we've put all the metadata files in place.

For sure. I shall remember.

**Pull request merged**